apps/full-site-editing/full-site-editing-plugin/starter-page-templates/class-wp-rest-image-sideload-controller.php

```diff
@@ -0,0 +1,96 @@
+<?php
+/**
+ * WP_REST_Image_Sideload_Controller file.
+ *
+ * @package full-site-editing
+ */
+
+/**
+ * Class WP_REST_Image_Sideload_Controller.
+ */
+class WP_REST_Image_Sideload_Controller extends WP_REST_Attachments_Controller {
+
+	/**
+	 * WP_REST_Image_Upload_Controller constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'wp/v2';
+		$this->rest_base = 'media/image';
+	}
+
+	/**
+	 * Register available routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => [ $this, 'create_item' ],
+					'permission_callback' => [ $this, 'create_item_permissions_check' ],
+					'show_in_index'       => false,
+				],
+				'schema' => [ $this, 'get_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Creates a single attachment.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|WP_REST_Response Response object on success, WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
+		if ( ! empty( $request['post'] ) && in_array( get_post_type( $request['post'] ), array( 'revision', 'attachment' ), true ) ) {
+			return new WP_Error( 'rest_invalid_param', __( 'Invalid parent type.' ), array( 'status' => 400 ) );
+		}
+
+		$url = esc_url_raw( $request->get_body() );
+
+		// Include image functions to get access to wp_read_image_metadata().
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+		require_once ABSPATH . 'wp-admin/includes/image.php';
+		require_once ABSPATH . 'wp-admin/includes/media.php';
+
+		$id = media_sideload_image( $url, 0, null, 'id' );
```

Comment:
This felt odd to me. I saw `$id` and immediately thought `int`, but it turns out it could be a `WP_Error` object. This felt confusing.

Is there a better var name to help make this more self-documenting?

`$sideload_result`?